crypto/provider_core.c

```diff
@@ -231,6 +224,9 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
         if (ref == 0) {
             DSO_free(prov->module);
             OPENSSL_free(prov->name);
+#ifndef HAVE_ATOMICS
+            CRYPTO_THREAD_lock_free(prov->refcnt_lock);
+#endif
```

Comment:
You could also make this unconditional

Reply:
I'll have a closer look at the HAVE_ATOMICS stuff.  We could do it better

Reply:
I planned on producing a macro or two that do the necessary conditioning for us.
It shouldn't be hard to make this a lot easier.